File: Iteration2-Python/src/Course.py
from typing import List
from abc import ABC,abstractmethod

import logging
logger = logging.getLogger(__name__)    # to ensure that the logger is named after the module

class Course(ABC):

#Constructor should be named  __init__
#don't need to declare or initialize attributes before the constructor, Attributes are dynamically created and initialized within the constructor

    def __init__(self, course_id: int, course_name: str, course_credits: int):
        self.__course_id = course_id
        self.__course_name = course_name
        self.__course_credits = course_credits
        self.__course_sections: List = []
        self.__prerequisites: List[Course] = []
        self.__student_list: List = []
        self.__waiting_student_list: List = []
        self.__course_capacity: int = 0  # TODO: Implement as per course sections
        self.number_of_lecture_hours: int = 0 #in the old java they were not private

    @abstractmethod
    def check_prerequisites_met(self, student) -> bool: #abstract
        pass
    
    def add_student_to_wait_list(self):
        pass

    def get_id(self) -> int:
        return self.__course_id

    def set_id(self, course_id: int):
        self.__course_id =course_id

    def get_credits(self) -> int:
        return self.__course_credits

    def set_credits(self, course_credits: int):
        self.__course_credits = course_credits

    def get_name(self) -> str:
        return self.__course_name

    def get_course_sections(self) -> List:
        return self.__course_sections

    def set_course_sections(self, course_sections: List):
        self.__course_sections = course_sections

    def add_course_section(self, section):
        self.__course_sections.append(section)

    def get_prerequisites(self) -> List:
        return self.__prerequisites

    def add_prerequisite(self, course):
        self.__prerequisites.append(course)

    def get_course_capacity(self) -> int:
        return sum(course_section.get_course_capacity() for course_section in self.get_course_sections())

File: Iteration1-Java/src/main/java/cse3063group3/Course.java
package cse3063group3;

import java.util.ArrayList;


public class Course {

    private int id;
    private String name;
    private int credits;
//    private int creditsNeeded;
    private ArrayList<Course> prerequisites;
    private ArrayList<Student> studentList;     // TODO: Send this and related methods into CourseSections.
    private ArrayList<CourseSection> courseSections;
    private ArrayList<Student> waitingStudentList = new ArrayList<>();
    int courseCapacity;                         // TODO: ^^^ Same as above.
    int numberOfLectureHours;

    public Course(int course_Id, String course_Name, int course_Credits) {
        this.id = course_Id;
        this.name = course_Name;
        this.credits = course_Credits;
        this.courseSections = new ArrayList<CourseSection>();
        this.prerequisites = new ArrayList<Course>();
    }

    /**
     * Checks if prerequisets for the current course are met or not by comparing argument studentFinishedCourseList to prerequisites.
     * @param studentFinishedCourseList An ArrayList containing a list of courses to compare against.
     * @return True if Course.ArrayList<Course> prerequisites is mett.<p>False otherwise.
     */
    public boolean CheckPrerequisitesMet(ArrayList<Course> studentFinishedCourseList, Student student){
        if (!studentFinishedCourseList.containsAll(this.prerequisites)) {
            //TODO: Send to controller.
            System.out.println("Did not finish all required courses to take " + this.getName() + ".");
            return false;
        }
        return true;
    }

    public void addStudentToWaitList() {
        // TODO: implement it
    }

    public int getId() {
        return this.id;
    }

    public void setId(int course_Id) {
        this.id = course_Id;
    }

    public int getCredits() {
        return this.credits;
    }

    public void setCredits(int course_Credits) {
        this.credits = course_Credits;
    }

    public String getName() {
        return this.name;
    }

    public ArrayList<CourseSection> getCourseSections() {
        return this.courseSections;
    }

    public void setCourseSection(ArrayList<CourseSection> course_Section_T) {
        this.courseSections = course_Section_T;
    }
    public void addCourseSection(CourseSection section){
        this.courseSections.add(section);
    }

    public ArrayList<Course> getPrerequisites() {
        return this.prerequisites;
    }

    public void setPrerequisites(Course course) {
        this.prerequisites.add(course);
    }

    public int getCourseCapacity() {
        int sum = 0;
        for(CourseSection courseSection : this.getCourseSections()){
            sum += courseSection.getCourseCapacity();
        }
        return sum;
    }
}
